### ISSUE TITLE
_history endpoint returning duplicate records 

### ISSUE THREAD

**Issue by pimrie-jembi** [hapifhir/hapi-fhir#3086]
**Describe the bug**
We send a bundle of resources to HAPI FHIR (in this example it is a bundle of 500 patients, each of them unique).

If we then run a _history call with paging (and count set to either 20 or 50 results per page : _history?_count=50 or _history?_count=20) we get duplicated patient records. It seems records from the first page of results are duplicated towards the end of the result set. 

We are subsequently losing resources because the duplicates replace valid records (so in this instance we get 500 results but 20 are duplicated which means only 480 unique resources are returned - instead of 500).

**To Reproduce**
1. Send a bundle of 500 unique resources to HAPI FHIR server
2. Using the _history endpoint retrieve the resources with paging enabled and count set to either 20 or 50
3. Check the result set for duplicate records

**Expected behavior**
500 unique records should be returned from the _history endpoint call

**Environment (please complete the following information):**
 - HAPI FHIR 5.4.2 - Docker
 - Ubuntu 20.04 LTS

**Comment by alexanderkiel** [comment]
We see the same issue.

**Comment by smailliwcs** [comment]
Seems to affect the latest JPA server.  See the following repro that posts a transaction bundle of 1000 resources, then counts unique history entries, paged ten at a time:

```sh
#!/bin/bash

docker run -p 8080:8080 -d hapiproject/hapi:latest
sleep 30

# Post a transaction bundle of 1000 resources
jq -n '{
  resourceType: "Bundle",
  type: "transaction",
  entry: [
    range(1000)
    | {
      fullUrl: "http://localhost:8080/fhir/Patient/\(.)",
      resource: {
        resourceType: "Patient"
      },
      request: {
        method: "POST",
        url: "Patient"
      }
    }
  ]
}' \
| curl -sSX POST 'http://localhost:8080/fhir' \
    -H 'Content-Type: application/fhir+json' \
    --data-binary @- \
    -o /dev/null

# Count unique history entries, paged ten at a time
url="http://localhost:8080/fhir/Patient/_history?_count=10&_format=json"
while true; do
    page="$(curl -sSX GET "$url")"
    if [ "$(echo "$page" | jq '.entry | length')" -eq 0 ]; then
        break
    fi
    echo "$page" | jq -r '.entry[].request | "\(.method) \(.url)"'
    url="$(echo "$page" | jq -r '.link[] | select(.relation == "next") | .url')"
    if [ -z "$url" ]; then
        break;
    fi
done \
| sort \
| uniq -c \
| awk '{ print $1; }' \
| sort \
| uniq -c
```

Typical output:

```
    763 1
    105 2
      9 3
```

Translation: among 1000 history entries, 763 appeared once, 105 appeared twice, and nine appeared three times (763 * 1 + 105 * 2 + 9 * 3 = 1000).  So only 763 + 105 + 9 = 877 unique entries appeared in the output, meaning that 123 were missing.

Note that this only occurs for paged history requests with resources initially created in a transaction bundle.  (So if you instead post a batch bundle, or if you request the entire history in one page, you get 1000 unique entries as expected.)  Presumably this is an issue with all those `meta.lastUpdated` elements being identical?

**Comment by smailliwcs** [comment]
Thanks for the fix!  Any thoughts on a workaround while waiting for a release (and then migrating to it)?  We're currently proving out an approach that overrides the `HistoryBuilder` bean with a `CustomHistoryBuilder` class that's basically a copy/paste of the upstream (HAPI) `HistoryBuilder`, but with the one-line fix from #5720.  Seems to work, but I'd be interested in a less crude alternative.
